File: Python/54. Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if matrix == []:
            return []
        ans = list()
        n, m = len(matrix), len(matrix[0])
        re, rs, ce, cs = n-1, 0, m-1, 0
        while True:
            print(cs, ce, rs, re)
            for i in range(cs, ce+1):
                ans.append(matrix[rs][i])
            rs += 1
            if rs >re:
                break
            for i in range(rs, re+1):
                ans.append(matrix[i][ce])
            ce -= 1
            if ce < cs:
                break
            for i in range(ce, cs-1, -1):
                ans.append(matrix[re][i])
            re -= 1
            if re < rs:
                break
            for i in range(re, rs-1, -1):
                ans.append(matrix[i][cs])
            cs += 1
            if cs > ce:
                break
        return ans

File: C++/54. Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;
        if(matrix.size() == 0)
            return ans;
        int rs = 0, cs = 0;
        int re = matrix.size()-1, ce = matrix[0].size() - 1;
        while(true)
        {
            for(int i=cs;i<=ce;i++)
                ans.push_back(matrix[rs][i]);
            rs++;
            if(rs > re) break;
            for(int i=rs;i<=re;i++)
                ans.push_back(matrix[i][ce]);
            ce--;
            if(ce < cs) break;
            for(int i=ce;i>=cs;i--)
                ans.push_back(matrix[re][i]);
            re--;
            if(re < rs) break;
            for(int i=re;i>=rs;i--)
                ans.push_back(matrix[i][cs]);
            cs++;
            if(cs>ce)
                break;
        }
        return ans;
    }
};